internal/cortex/frontend/transport/handler.go

```diff
@@ -55,20 +58,40 @@ type Handler struct {
 	cfg          HandlerConfig
 	log          log.Logger
 	roundTripper http.RoundTripper
+	lruCache     *lru.Cache
+	regex        *regexp.Regexp
+	errorExtract *regexp.Regexp
 
 	// Metrics.
 	querySeconds *prometheus.CounterVec
 	querySeries  *prometheus.CounterVec
 	queryBytes   *prometheus.CounterVec
+	cachedHits   prometheus.Counter
 	activeUsers  *util.ActiveUsersCleanupService
 }
 
 // NewHandler creates a new frontend handler.
 func NewHandler(cfg HandlerConfig, roundTripper http.RoundTripper, log log.Logger, reg prometheus.Registerer) http.Handler {
+	var (
+		LruCache *lru.Cache
+		err      error
+	)
+
+	if cfg.FailedQueryCacheCapacity > 0 {
+		LruCache, err = lru.New(cfg.FailedQueryCacheCapacity)
+		if err != nil {
+			LruCache = nil
+			level.Warn(log).Log("msg", "Failed to create LruCache", "error", err)
+		}
+	}
+
 	h := &Handler{
 		cfg:          cfg,
 		log:          log,
 		roundTripper: roundTripper,
+		lruCache:     LruCache,
+		regex:        regexp.MustCompile(`[\s\n\t]+`),
+		errorExtract: regexp.MustCompile(`Code\((\d+)\)`),
```

Comment:
does the error string hand-picked from m3 response?

Reply:
I looked at the err.error() string and then pattern matched on the error code present in that message. @hczhu-db